src/compiler/checker.ts

```diff
@@ -26299,7 +26300,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             if (propName === undefined) {
                 return type;
             }
-            const removeNullable = strictNullChecks && isOptionalChain(access) && maybeTypeOfKind(type, TypeFlags.Nullable);
+            const removeNullable = strictNullChecks && (isOptionalChain(access) || isNonNullAccess(access)) && maybeTypeOfKind(type, TypeFlags.Nullable);

```

Comment:
Here we remove `null` or `undefined` from the type, but then a few lines below we add back only an `undefined`. That reflects what happens in an optional chain, but the `!` operator never changes one to the other. It may not matter, but doesn't quite seem right.

Reply:
I see, I had missed that. So I guess we shouldn't call `getOptionalType` below if it's a `!` operator, since `!` actually gets rid of `undefined` and `null`, unlike an optional chain.

Reply:
@ahejlsberg I pushed the above change, let me know if it's not right or if there's something else missing.